src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -0,0 +1,666 @@
+
+  for (auto& payment: payments) {
+    std::string address_str = cryptonote::get_account_address_as_str(m_nettype, 0, payment.address_info.address);
+    MTRACE("Adding record for SN reward contributor " << address_str << "to database with amount " << static_cast<int64_t>(payment.amount));
+
+    db::exec_query(insert_payment, address_str, static_cast<int64_t>(payment.amount * 1000));
+    insert_payment.reset();
+  };
+
+  return true;
+}
+
+bool BlockchainSQLite::subtract_sn_payments(std::vector<cryptonote::batch_sn_payment>& payments)
+{
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
+	SQLite::Statement update_payment{db,
+		"UPDATE batched_payments_accrued SET amount = (amount - ?) WHERE address = ?"};
+
+	for (auto& payment: payments) {
+		std::string address_str = cryptonote::get_account_address_as_str(m_nettype, 0, payment.address_info.address);
+		auto result = db::exec_query(update_payment, static_cast<int64_t>(payment.amount * 1000), address_str);
+    if (!result) {
+      MERROR("tried to subtract payment from an address that doesnt exist: " << address_str);
+      return false;
+    }
+		update_payment.reset();
+	};
+
+  return true;
+}
+
+std::optional<std::vector<cryptonote::batch_sn_payment>> BlockchainSQLite::get_sn_payments(uint64_t block_height)
+{
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
+
+  if (block_height == 0)
+    return std::nullopt;
+
+  const auto& conf = get_config(m_nettype);
+
+  SQLite::Statement select_payments{db,
+    "SELECT address, amount FROM batched_payments_accrued WHERE amount > ? ORDER BY address ASC"}; 
+
+  select_payments.bind(1, static_cast<int64_t>(conf.MIN_BATCH_PAYMENT_AMOUNT));
+
+  std::vector<cryptonote::batch_sn_payment> payments;
+
+  std::string address;
+  uint64_t amount;
+  while (select_payments.executeStep())
+  {
+    address = select_payments.getColumn(0).getString();
+    amount = static_cast<uint64_t>(select_payments.getColumn(1).getInt64() / 1000);
+    if (cryptonote::is_valid_address(address, m_nettype)) {
+      cryptonote::address_parse_info addr_info{};
+      cryptonote::get_account_address_from_str(addr_info, m_nettype, address);
+      uint64_t next_payout_height = addr_info.address.next_payout_height(block_height - 1, conf.BATCHING_INTERVAL);
+      if (block_height == next_payout_height)
+      {
+        payments.emplace_back(address, amount, m_nettype);
+      }
+    }
+    else
+    {
+      MERROR("Invalid address returned from batching database: " << address);
+      return std::nullopt;
+    }
+  }
+
+  return payments;
+}
+
+std::vector<cryptonote::batch_sn_payment> BlockchainSQLite::calculate_rewards(uint8_t hf_version, uint64_t distribution_amount, service_nodes::service_node_info sn_info)
+{
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
+
+  // Find out how much is due for the operator
+  uint64_t operator_fee = 0;
+  {
```

Comment:
should be its own function; either in anonymous namespace here or elsewhere if this logic is used elsewhere.